### PR TITLE
R connector for ADLS

### PR DESCRIPTION
**Pull request by omkarksa**
Implementation of an R connector for ADLS, which is based on the R connector for Azure Blob. The code is complete with a small test suite.

### CODE REVIEW

File: R/AzureDataLake.R [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+#'
+#' @inheritParams createAzureContext
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
+#' @param relativePath Provide a relative path of the directory.
+#' @param verbose Print tracing information (default FALSE).
+#'
+#' @return Returns a data frame.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: But what's _in_ the data frame? One row per directory? What are the columns? And error conditions?

File: R/AzureDataLake.R [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+azureDataLakeListStatus <- function(azureActiveContext, azureDataLakeAccount, relativePath = "", verbose = FALSE) {
+
+  if (!missing(azureActiveContext) && !is.null(azureActiveContext)) {
+    assert_that(is.azureActiveContext(azureActiveContext))
+    azureCheckToken(azureActiveContext)
+  }
+  assert_that(is_storage_account(azureDataLakeAccount))
+  verbosity <- set_verbosity(verbose)
+
+  URL <- paste0(
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: Can all these repeated string literals be centralized?

File: R/AzureDataLake.R [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+#'
+#' @inheritParams createAzureContext
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
+#' @param relativePath Provide a relative path of the directory.
+#' @param verbose Print tracing information (default FALSE).
+#'
+#' @return Returns a data frame.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: Say more? In fact, sometimes it returns null.

File: R/AzureDataLake.R [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    relativePath,
+    "?op=GETFILESTATUS",
+    "&api-version=2016-11-01"
+  )
+  
+  resHttp <- callAzureDataLakeApi(URL,
+                                  azureActiveContext = azureActiveContext,
+                                  verbose = verbose)
+  
+  if (status_code(resHttp) == 404) {
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: Error handling seems inconsistent and arbitrary. If this is worth doing here, isn't it worth doing for azureDataLakeListStatus ?

File: R/AzureDataLake.R [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    relativePath,
+    "?op=LISTSTATUS",
+    "&api-version=2016-11-01"
+    )
+
+  resHttp <- callAzureDataLakeApi(URL,
+    azureActiveContext = azureActiveContext,
+    verbose = verbose)
+
+  stopWithAzureError(resHttp)
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: This seems extreme.

File: R/AzureDataLake.R [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+#' @inheritParams createAzureContext
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
+#' @param relativePath Provide a relative path of the directory.
+#' @param permission Provide the permission to be set for the directory.
+#' @param verbose Print tracing information (default FALSE).
+#'
+#' @return Returns a boolean.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: Indicating ...

File: R/AzureDataLake.R [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+#' Azure Data Lake CREATE for specified relativePath of an azure data lake account.
+#'
+#' @inheritParams createAzureContext
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
+#' @param relativePath Provide a relative path of the directory.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: File?

File: R/AzureDataLake.R [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+#' @param overwrite Overwrite existing files (default FALSE).
+#' @param permission Provide the permission to be set for the directory.
+#' @param contents Provide contents to write to `relativePath`
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: Any constraints on the contents? Type? Size? Are they checked in the code? Tested?

File: R/AzureDataLake.R [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+  )
+  if (!missing(overwrite)  && !is.null(overwrite)) URL <- paste0(URL, "&overwrite=", overwrite)
+  if (!missing(permission)  && !is.null(permission)) URL <- paste0(URL, "&permission=", permission)
+  
+  resHttp <- callAzureDataLakeApi(URL, verb = "PUT",
+                                  azureActiveContext = azureActiveContext,
+                                  content = contents, contenttype = "text/plain; charset=UTF-8",
+                                  verbose = verbose)
+  stopWithAzureError(resHttp)
+  # TODO: Check why this returns NULL
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: I think it's an error handling issue -- the error handling is too brutal ...

File: R/AzureDataLake.R [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+#' Azure Data Lake APPEND for specified relativePath of an azure data lake account.
+#'
+#' @inheritParams createAzureContext
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
+#' @param relativePath Provide a relative path of the directory.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: File?

File: R/AzureDataLake.R [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+  # TODO: Check why this returns NULL
+}
+
+#' Azure Data Lake APPEND for specified relativePath of an azure data lake account.
+#'
+#' @inheritParams createAzureContext
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: As before?

File: R/AzureDataLake.R [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+    "?op=APPEND&append=true",
+    "&api-version=2016-11-01"
+  )
+  
+  resHttp <- callAzureDataLakeApi(URL, verb = "POST",
+                                  azureActiveContext = azureActiveContext,
+                                  content = contents, contenttype = "text/plain; charset=UTF-8",
+                                  verbose = verbose)
+  stopWithAzureError(resHttp)
+  # TODO: Check why this returns NULL
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: Ditto.

File: R/AzureDataLake.R [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
+#' @param relativePath Provide a relative path of the directory.
+#' @param offset Provide the offset to read from.
+#' @param length Provide length of data to read.
+#' @param verbose Print tracing information (default FALSE).
+#'
+#' @return Returns a data frame.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: What's in it? The data read, obviously, but in what format?

File: R/AzureDataLake.R [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+#' @inheritParams createAzureContext
+#' @inheritParams azureAuthenticate
+#'
+#' @param azureActiveContext Provide an `azureActiveContext` object used for authentication.
+#' @param azureDataLakeAccount Provide the name of the Azure Data Lake account.
+#' @param relativePath Provide a relative path of the directory.
+#' @param recursive Provide recursive delete option.
+#' @param verbose Print tracing information (default FALSE).
+#'
+#' @return Returns a boolean.
```
Comment (spirom-microsoft) [inline R/AzureDataLake.R]: Indicating whether it deleted, or something else?

File: tests/testthat/test-7-datalake.R [hunk 15]
```diff
@@ -0,0 +1,91 @@
+
```
Comment (spirom-microsoft) [inline tests/testthat/test-7-datalake.R]: I made some general comments on the tests in my other review.

File: R/AzureDataLake.R (hunk 5 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: This stops only incase of errors.

File: R/AzureDataLake.R (hunk 1 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 5 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 3 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 4 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 6 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 7 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 8 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 9 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Designed this way.

File: R/AzureDataLake.R (hunk 11 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 10 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 12 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Designed this way.

File: R/AzureDataLake.R (hunk 13 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: R/AzureDataLake.R (hunk 14 above)
Comment (omkarksa) [reply R/AzureDataLake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

File: tests/testthat/test-7-datalake.R (hunk 15 above)
Comment (omkarksa) [reply tests/testthat/test-7-datalake.R]: Fixed in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

**Comment by omkarksa** [comment]
Fixed all issues in https://github.com/omkarksa/AzureSMR/pull/1 will raise a pull request once this is merged into master

**Pull request merged**